### PR TITLE
Optionally nocanon

### PR DESCRIPTION
**Pull request by mariodsantana**
[Resubmission of pull #961 after rebasing for more alternatives.]

These two small patches fix issues I'm having when trying to do some tricky proxy actions. For example if I do:

SecDefaultAction "phase:2,log,setvar:SESSION.badness=1,proxy:'http://192.168.190.12%{REQUEST_URI_RAW}"

I've been applying them to new releases without issue for my use case. I posted them a while back as "issues" but they never got answered there? Submitting as a pull request now in case that's the preferred method. The fixes are:
1. when executing disruptive actions, it's necessary to execute the intercept action as well. I don't think this will break anything, but it certainly fixes this scenario where mod_proxy doesn't expand the REQUEST_URI_RAW macro. Maybe we should only do this for proxy actions?
2. give the user the option not to "canonicalize" the proxy URL - if we want to send the URI unchanged to the backend server to whom we are proxying, specify "[nocanon]" as the first characters in the target URL.

### CODE REVIEW

File: apache2/re.c [hunk 1]
```diff
@@ -2513,6 +2513,11 @@ static void msre_perform_disruptive_actions(modsec_rec *msr, msre_rule *rule,
             }
         }
     }
+    if (actionset->intercept_action_rec->metadata->type == ACTION_DISRUPTIVE) {
```
Comment (zimmerle) [inline apache2/re.c]: "proxy-nocanon" should be checked here, right?

File: tests/regression/action/00-disruptive-actions.t [hunk 2]
```diff
@@ -476,6 +476,37 @@
 },
 {
 	type => "action",
+	comment => "nocanon proxy in phase:1 (get)",
```
Comment (zimmerle) [inline tests/regression/action/00-disruptive-actions.t]: Great! Those test cases are very helpful. ;)

**Comment by zimmerle** [comment]
Hi @mariodsantana,

Thank you for the patch. I made some comments inline in the code.

**Comment by uxbod** [comment]
Hello and a Happy New Year!

Am running NGINX 1.9.9 with the nginx_refactored branch in a proxy configuration.  For some reason, most probably configuration, the following rule never appears to get applied:

SecRule REMOTE_ADDR "@pmFromFile /etc/asl/whitelist" phase:1,pass,t:none,id:318743,nolog,skipAfter:END_RBL

Even though I have added the IP to the whitelist file it is still hitting this rule:

SecRule REQUEST_HEADERS:Host "^[\d.:]+$" \
"chain,phase:2,rev:2,t:none,pass,msg:'Atomicorp.com WAF Rules: Suspicious activity detected - Host header is a numeric IP address', severity:'2',id:'331032',severity:'5',tag:'no_ar'"
SecRule REMOTE_ADDR "!@ipMatch 127.0.0.1,::1"  "t:none"

Any thoughts as to why its being missed in phase 1 ?

Thanks, Phil

File: apache2/re.c (hunk 1 above)
Comment (mariodsantana) [inline apache2/re.c]: I don't think so...?  Here's my understanding:

The loop before this line looks through all the actions in `actionset->actions` and run the execute callback for each.  For some reason, though, it doesn't seem like the `actionset->intercept_action_rec` is part of the list in `actionset->actions` -- so I added a few lines to check whether `actionset->intercept_action_rec` is an intercept action, and if so to run the execute callback for it.

File: tests/regression/action/00-disruptive-actions.t (hunk 2 above)
Comment (mariodsantana) [inline tests/regression/action/00-disruptive-actions.t]: Just trying to be a good citizen.  I appreciate all the work you guys do!  :)

**Comment by mariodsantana** [comment]
Thanks for the feedback, @zimmerle - I've responded to your comments.

File: apache2/re.c (hunk 1 above)
Comment (zimmerle) [inline apache2/re.c]: I've got your point. You are right.

File: tests/regression/action/00-disruptive-actions.t (hunk 2 above)
Comment (zimmerle) [inline tests/regression/action/00-disruptive-actions.t]: Thanks ;)

**Comment by zimmerle** [comment]
Hi @mariodsantana,

Thank you again for you path. 

Your patch is now on our buildbots. Made a small modification, as you can see here:
https://github.com/SpiderLabs/ModSecurity/commit/c711808ef73a16df830cebf3e381846510942ed3

**Comment by zimmerle** [comment]
Documentation was changed already: https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual#proxy

**Comment by zimmerle** [comment]
Merged! It will be released as part of ModSecurity v2.9.1

**Comment by mariodsantana** [comment]
Thanks!

**Pull request closed without merging**